### PR TITLE
DAOS-11551 control: Flush c output buffers.

### PR DESCRIPTION
**Pull request by ashleypittman**
Ensure test written from C subroutines in go is printed.

Required-githooks: true

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

File: src/control/cmd/daos/util.go [hunk 1]
```diff
@@ -153,33 +152,36 @@ func createWriteStream(ctx context.Context, prefix string, printLn func(line str
...
+			if prefix == "" {
+				printLn(fmt.Sprintf("%s", line))
+			} else {
+				printLn(fmt.Sprintf("%s: %s", prefix, line))
 			}
 		}
 	}(ctx, prefix)
 
 	return stream, func() {
+		w.WriteString("close\n")
```
Comment (daltonbohning) [inline src/control/cmd/daos/util.go:179]: This seems like a workaround. Is there not a more direct approach?

File: src/control/cmd/daos/util.go [hunk 2]
```diff
@@ -153,33 +152,36 @@ func createWriteStream(ctx context.Context, prefix string, printLn func(line str
...
+		w.Sync()
+		C.fflush(stream)
+		C.fflush(stream)
 		C.fflush(stream)
```
Comment (daltonbohning) [inline src/control/cmd/daos/util.go:183]: 3 flushes on the same stream?

Comment (ashleypittman) [reply src/control/cmd/daos/util.go:183]: Old School.  Will revise.

File: src/control/cmd/daos/util.go (hunk 1 above)
Comment (ashleypittman) [reply src/control/cmd/daos/util.go:180]: Perhaps.  There are two things going on here, getting the consumer not the producer to shut the pipe but also to ensure that the pipe is not just flushed but also empty, and the only way I know how to do that is to send some kind of tag through it.

Comment (tanabarr) [reply src/control/cmd/daos/util.go:180]: this seems okay as a temporary workaround IMO, we can follow-up with a more concrete fix

**Review by tanabarr — APPROVED** [review]

File: src/control/cmd/daos/util.go [hunk 3]
```diff
@@ -153,33 +152,33 @@ func createWriteStream(ctx context.Context, prefix string, printLn func(line str
 	}
 
 	go func(ctx context.Context, prefix string) {
-		if prefix != "" {
-			prefix = ": "
-		}
```
Comment (ashleypittman) [inline src/control/cmd/daos/util.go:158]: This is also a (different) bug fix, this code was obviously wrong.

File: src/control/cmd/daos/util.go (hunk 1 above)
Comment (ashleypittman) [reply src/control/cmd/daos/util.go:180]: I don't know what a more concrete fix would be, even using a semaphore to message from the producer to the consumer that it should exit there's still no way to flush a buffered pipe that I'm aware of.

**Comment by ashleypittman** [comment]
I wasn't very happy with this so I've tweaked it a bit further.

File: src/control/cmd/daos/util.go [hunk 4]
```diff
@@ -152,33 +150,24 @@ func createWriteStream(ctx context.Context, prefix string, printLn func(line str
...
+	go func(ctx context.Context) {
 
 		rdr := bufio.NewReader(r)
 		for {
-			select {
-			case <-ctx.Done():
-				return
```
Comment (kjacque) [inline src/control/cmd/daos/util.go:164]: I'm not sure about removing this. Is this what's cutting off the output? The purpose of these lines is to exit this goroutine in the event that the user exits the program.

File: src/control/cmd/daos/util.go [hunk 5]
```diff
@@ -152,33 +150,24 @@ func createWriteStream(ctx context.Context, prefix string, printLn func(line str
...
+				r.Close()
+				return
 			}
+			printLn(line)
 		}
-	}(ctx, prefix)
+	}(ctx)
 
 	return stream, func() {
-		C.fflush(stream)
```
Comment (kjacque) [inline src/control/cmd/daos/util.go:179]: Don't need to force-flush anymore?

File: src/control/cmd/daos/util.go (hunk 4 above)
Comment (ashleypittman) [reply src/control/cmd/daos/util.go:164]: Will try re-adding, thank you for the review.

File: src/control/cmd/daos/util.go (hunk 5 above)
Comment (ashleypittman) [reply src/control/cmd/daos/util.go:179]: No.

**Review by tanabarr — APPROVED** [review]

**Review by kjacque — APPROVED** [review]

**Comment by ashleypittman** [comment]
There's a typo in the commit message and I'm not at my laptop to fix it now.

**Comment by ashleypittman** [comment]
Do not land this, further work needed.

**Comment by ashleypittman** [comment]
This only partially worked so dropping in favour of https://github.com/daos-stack/daos/pull/12632 which is similar but uses a channel to communicate back success.

**Pull request closed without merging**